### PR TITLE
Group chapter progress: group / teams version

### LINKED ISSUE #224: Group chapter progress: group / teams version
## Motivations
Implement the group and team tabs as shown in the design.

## Subtasks
* Use the dedicated services (see API) to implement the same matrix as users but for teams and groups.

### PR DESCRIPTION
**Pull request by GabrielVidal1**
Fixes #224

When fetching the progress for sub-groups (using this service [Get group progress](https://france-ioi.github.io/algorea-devdoc/api/#operation/groupGroupProgress)) I get a `validation_rate` variable, which is a float between 0 and 1 (and not a boolean), so what is the condition to say that a group validated an item?
For now I consider the item is validated if the `validation_rate` equals 1.

### CODE REVIEW

**Review by smadbe — CHANGES_REQUESTED** [review]
First quick UI check: on first arriving  on the page, there is not spinner for loading the user table.
I will try next week to find some data, because it is empty and cannot check anything like that (I don't know how you do actually).

File: src/app/modules/item/components/item-chapter-view/group-situation-chapter-view/group-situation-chapter-view.component.html [hunk 1]
```diff
@@ -1,10 +1,27 @@
...
 <ng-template #noError>
+
+  <div class="loader">
+    <p-progressSpinner *ngIf="state === 'fetching' && data.rows.length === 0"></p-progressSpinner>
+  </div>
+
+  <ng-container *ngIf="state === 'ready' && data.rows.length === 0" [ngSwitch]="data.type">
+    <p class="empty-message" *ngSwitchCase="'Users'">This group has no users</p>
+    <p class="empty-message" *ngSwitchCase="'Teams'">This group has no teams</p>
+    <p class="empty-message" *ngSwitchCase="'Groups'">This group has no groups</p>
```
Comment (smadbe) [inline src/app/modules/item/components/item-chapter-view/group-situation-chapter-view/group-situation-chapter-view.component.html:20]:
```suggestion
    <p class="empty-message" *ngSwitchCase="'Groups'">This group has no sub-groups</p>
```

File: src/app/modules/item/components/item-chapter-view/group-situation-chapter-view/group-situation-chapter-view.component.html [hunk 2]
```diff
@@ -1,10 +1,27 @@
...
+
 <p *ngIf="state === 'error';else noError" class="error-message">
   <i class="fa fa-exclamation-triangle"></i>
   Error while loading the data
 </p>
 
 <ng-template #noError>
+
+  <div class="loader">
+    <p-progressSpinner *ngIf="state === 'fetching' && data.rows.length === 0"></p-progressSpinner>
```
Comment (smadbe) [inline src/app/modules/item/components/item-chapter-view/group-situation-chapter-view/group-situation-chapter-view.component.html:14]: shouldn't the ngIf be moved to the div?

File: src/app/modules/item/components/item-chapter-view/group-situation-chapter-view/group-situation-chapter-view.component.ts [hunk 3]
```diff
@@ -2,17 +2,28 @@ import { Component, Input, OnChanges, OnDestroy, SimpleChanges } from '@angular/
...
+  itemId: string,
+  validated: boolean,
+  score: number,
+  timeSpent: number,
+}
 
 interface Data {
-  users: FormattableUser[],
+  type: TypeFilter,
+  rows: string[],
```
Comment (smadbe) [inline src/app/modules/item/components/item-chapter-view/group-situation-chapter-view/group-situation-chapter-view.component.ts:24]: `rowHeaders`maybe?

File: src/app/shared/pipes/userDisplay.ts [hunk 4]
```diff
@@ -6,17 +6,19 @@ export interface FormattableUser {
...
+    return `${userInfos.firstName} ${userInfos.lastName} (${userInfos.login})`;
+  } else if (userInfos.firstName !== null && userInfos.firstName !== '') {
+    return `${userInfos.firstName} (${userInfos.login})`;
+  } else if (userInfos.lastName !== null && userInfos.lastName !== '') {
+    return `${userInfos.lastName} (${userInfos.login})`;
+  } else {
+    return userInfos.login;
+  }
+}
+
```
Comment (smadbe) [inline src/app/shared/pipes/userDisplay.ts:20]: I would prefer you move this function out of the pipe. For instance, in `helpers`.

**Review by smadbe — CHANGES_REQUESTED** [review]
👍 very good. Slight changes only.

File: src/app/modules/item/components/item-chapter-view/group-situation-chapter-view/group-situation-chapter-view.component.html (hunk 1 above)
Comment (GabrielVidal1) [reply src/app/modules/item/components/item-chapter-view/group-situation-chapter-view/group-situation-chapter-view.component.html]: fixed

File: src/app/modules/item/components/item-chapter-view/group-situation-chapter-view/group-situation-chapter-view.component.html (hunk 2 above)
Comment (GabrielVidal1) [reply src/app/modules/item/components/item-chapter-view/group-situation-chapter-view/group-situation-chapter-view.component.html]: fixed

File: src/app/modules/item/components/item-chapter-view/group-situation-chapter-view/group-situation-chapter-view.component.ts (hunk 3 above)
Comment (GabrielVidal1) [reply src/app/modules/item/components/item-chapter-view/group-situation-chapter-view/group-situation-chapter-view.component.ts]: fixed

File: src/app/shared/pipes/userDisplay.ts (hunk 4 above)
Comment (GabrielVidal1) [reply src/app/shared/pipes/userDisplay.ts]: fixed

**Review by smadbe — APPROVED** [review]

**Pull request merged**